<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function performs a MaAsLin2 analysis
(https://huttenhower.sph.harvard.edu/maaslin/), which finds associations
between microbiome meta-omics features and metadata. Features consist on the
abundance of gene families or metabolic pathways of multiple samples,
obtained using HUMAnN3 (https://huttenhower.sph.harvard.edu/humann)."><title>Run MaAsLin2 Analysis Quietly — run_maaslin2 • microfunk</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run MaAsLin2 Analysis Quietly — run_maaslin2"><meta property="og:description" content="This function performs a MaAsLin2 analysis
(https://huttenhower.sph.harvard.edu/maaslin/), which finds associations
between microbiome meta-omics features and metadata. Features consist on the
abundance of gene families or metabolic pathways of multiple samples,
obtained using HUMAnN3 (https://huttenhower.sph.harvard.edu/humann)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">microfunk</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MicrobialGenomics-IrsicaixaOrg/microfunk/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Run MaAsLin2 Analysis Quietly</h1>
      <small class="dont-index">Source: <a href="https://github.com/MicrobialGenomics-IrsicaixaOrg/microfunk/blob/HEAD/R/maaslin2.R" class="external-link"><code>R/maaslin2.R</code></a></small>
      <div class="d-none name"><code>run_maaslin2.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function performs a MaAsLin2 analysis
(https://huttenhower.sph.harvard.edu/maaslin/), which finds associations
between microbiome meta-omics features and metadata. Features consist on the
abundance of gene families or metabolic pathways of multiple samples,
obtained using HUMAnN3 (https://huttenhower.sph.harvard.edu/humann).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_maaslin2</span><span class="op">(</span></span>
<span>  <span class="va">se</span>,</span>
<span>  output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"/output_folder"</span><span class="op">)</span>,</span>
<span>  min_abundance <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min_prevalence <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  min_variance <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  normalization <span class="op">=</span> <span class="st">"NONE"</span>,</span>
<span>  transform <span class="op">=</span> <span class="st">"NONE"</span>,</span>
<span>  analysis_method <span class="op">=</span> <span class="st">"LM"</span>,</span>
<span>  max_significance <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  random_effects <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fixed_effects <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  correction <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  standardize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_heatmap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  heatmap_first_n <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  plot_scatter <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  max_pngs <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  save_scatter <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  save_models <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  reference <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>se</dt>
<dd><p>A SummarizedExperiment object containing normalized abundance
values as the assay and metadata as the column data.</p></dd>


<dt>output</dt>
<dd><p>Character string specifying the output directory path. Default
is a tmp folder.</p></dd>


<dt>min_abundance</dt>
<dd><p>Minimum abundance for each feature. Default is 0.0.</p></dd>


<dt>min_prevalence</dt>
<dd><p>Minimum percent of samples for which a feature is
detected at minimum abundance. Default is 0.1.</p></dd>


<dt>min_variance</dt>
<dd><p>Minimum variance threshold for features to be included.
Default is 0.0.</p></dd>


<dt>normalization</dt>
<dd><p>Normalization method to apply. Default is 'NONE'.</p></dd>


<dt>transform</dt>
<dd><p>Transformation method to be used. Default is 'NONE'.</p></dd>


<dt>analysis_method</dt>
<dd><p>Analysis method to apply. Default is 'LM'.</p></dd>


<dt>max_significance</dt>
<dd><p>Q-value threshold for significance. Default is 0.25.</p></dd>


<dt>random_effects</dt>
<dd><p>Random effects to be included in the analysis
(comma-delimited for multiple effects). Default is NULL.</p></dd>


<dt>fixed_effects</dt>
<dd><p>Fixed effects to be included in the analysis
(comma-delimited for multiple effects). Default is NULL.</p></dd>


<dt>correction</dt>
<dd><p>Correction method for computing the Q-value. Default is
'BH'.</p></dd>


<dt>standardize</dt>
<dd><p>Apply Z-score so continuous metadata are on the same
scale. Default is FALSE.</p></dd>


<dt>plot_heatmap</dt>
<dd><p>Generate a heatmap for significant associations. Default
is TRUE.</p></dd>


<dt>heatmap_first_n</dt>
<dd><p>In the heatmap, plot only top N features with
significant associations. Default is 50.</p></dd>


<dt>plot_scatter</dt>
<dd><p>Generate scatter plots for the significant associations.
Default is TRUE.</p></dd>


<dt>max_pngs</dt>
<dd><p>Set the maximum number of scatter plots for significant
associations to save as png files. Default is 10.</p></dd>


<dt>save_scatter</dt>
<dd><p>Save all scatter plot ggplot objects to an RData file.
Default is FALSE.</p></dd>


<dt>save_models</dt>
<dd><p>Return the full model outputs and save to an RData file.
Default is FALSE.</p></dd>


<dt>reference</dt>
<dd><p>Factor to use as reference for categorical variables,
provided as a string of 'variable,reference' (semi-colon delimited for
multiple variables).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Results of the MaAsLin2 analysis as a tibble.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function takes a SummarizedExperiment object containing normalized gene/
pathway abundance values as the assay and metadata as the column data, as
well as several parameter options to run the analysis. It creates a temporary
output folder to store created files and returns the result of the analysis
on the console.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Read HUMAnN3 &amp; MaAsLin2 Analysis</span></span></span>
<span class="r-in"><span><span class="va">da_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="read_humann.html">read_humann</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"reduced_genefam_cpm_kegg.tsv"</span>, package <span class="op">=</span> <span class="st">"microfunk"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  metadata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"reduced_meta.csv"</span>, package <span class="op">=</span> <span class="st">"microfunk"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> <span class="fu">run_maaslin2</span><span class="op">(</span>fixed_effects <span class="op">=</span> <span class="st">"ARM"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">da_result</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 3,806 × 11</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    feature metadata da_method coefficient log2FC stderr lfc_stderr p_value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> K03592  ARM      maaslin2       -<span style="color: #BB0000;">0.389</span> -           0 -                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> K03788  ARM      maaslin2       -<span style="color: #BB0000;">0.125</span> -           0 -                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> K05708  ARM      maaslin2       -<span style="color: #BB0000;">0.189</span> -           0 -                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> K07270  ARM      maaslin2       -<span style="color: #BB0000;">0.567</span> -           0 -                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> K08678  ARM      maaslin2       -<span style="color: #BB0000;">0.793</span> -           0 -                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> K13652  ARM      maaslin2       -<span style="color: #BB0000;">1.08</span>  -           0 -                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> K16950  ARM      maaslin2       -<span style="color: #BB0000;">1.71</span>  -           0 -                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> K20861  ARM      maaslin2       -<span style="color: #BB0000;">0.272</span> -           0 -                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> K21055  ARM      maaslin2       -<span style="color: #BB0000;">0.588</span> -           0 -                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> K21065  ARM      maaslin2       -<span style="color: #BB0000;">0.578</span> -           0 -                0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 3,796 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 3 more variables: q_value &lt;dbl&gt;, padj_value &lt;chr&gt;, signif &lt;lgl&gt;</span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Judit Farré-Badia, Francesc Català-Moll.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

